src/charm.py

```diff
@@ -1,103 +1,141 @@
+            return CharmState.from_charm(self)
+        except CharmConfigInvalidError as exc:
+            self.unit.status = ops.BlockedStatus(str(exc))
+            return None
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def _on_install(self, event: ops.InstallEvent) -> None:
+        """Handle installation of the charm.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        Installs apt packages required to build the image.
+
+        Args:
+            event: The event fired on install hook.
+        """
+        self.unit.status = ops.MaintenanceStatus("Setting up Builder.")
+        state = self._load_state()
+        if not state:
+            # Defer this event since on install should be re-triggered to setup dependencies for
+            # the charm. Since the charm goes into blocked state and the user reconfigures the
+            # state, config_changed event should be queued after the deferred on_install.
+            event.defer()
+            return
+
+        builder.configure_proxy(proxy=state.proxy_config)
+        builder.setup_builder()
+        self.unit.status = ops.WaitingStatus("Waiting for first image build.")
+
+    def _build_image(self, state: CharmState) -> str:
+        """Build image and propagate the new image.
+
+        Args:
+            state: The charm state.
+
+        Returns:
+            The built image ID.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+        self.unit.status = ops.StatusBase.from_name(self.unit.status.name, "Building image.")
+        build_config = builder.BuildImageConfig(
+            arch=state.image_config.arch, base_image=state.image_config.base_image
+        )
+        image_path = builder.build_image(config=build_config)
+        upload_config = openstack_manager.UploadImageConfig(
+            arch=state.image_config.arch,
+            app_name=self.app.name,
+            base=state.image_config.base_image,
+            num_revisions=state.revision_history_limit,
+            src_path=image_path,
+        )
+        self.unit.status = ops.StatusBase.from_name(self.unit.status.name, "Updating image.")
+        with openstack_manager.OpenstackManager(cloud_config=state.cloud_config) as openstack:
+            image_id = openstack.upload_image(config=upload_config)
+
+        self.image_observer.update_relation_data(image_id=image_id)
+        return image_id
```

Comment:
this should probably be a snap - let's move to that and then we can have another review 🙂